src/cloud-api-adaptor/pkg/securecomms/wnssh/wnssh.go

```diff
@@ -0,0 +1,386 @@
+			log.Printf("failed to create PP Secret: %v", err)
+			return nil
+		}
+	}
+
+	// >>> Update the KBS about the SID's Secret !!! <<<
+	sidSecretPath := fmt.Sprintf("default/pp-%s/privateKey", sid)
+	log.Printf("Updating KBS with secret for: %s", sidSecretPath)
+	err = c.kc.PostResource(sidSecretPath, privateKey)
+	if err != nil {
+		log.Printf("failed to PostResource PP Secret: %v", err)
+		return nil
+	}
+
+	var serverSshPublicKeyBytes []byte
+
+	if len(publicKey) > 0 {
+		serverSshPublicKey, _, _, _, err := ssh.ParseAuthorizedKey(publicKey)
+		if err != nil {
+			log.Printf("unable to ParseAuthorizedKey serverPublicKey: %v", err)
+			return nil
+		}
+		serverSshPublicKeyBytes = serverSshPublicKey.Marshal()
+	}
+
+	ppAddr := make([]string, len(ipAddr))
+	for i, ip := range ipAddr {
+		ppAddr[i] = ip.String() + ":" + sshutil.SSHPORT
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	ci := &SshClientInstance{
+		sid:          sid,
+		publicKey:    serverSshPublicKeyBytes,
+		ppAddr:       ppAddr,
+		sshClient:    c,
+		ctx:          ctx,
+		cancel:       cancel,
+		inboundPorts: make(map[string]string),
+	}
+
+	for _, tag := range c.inboundStrings {
+		name, inPort, err := ci.inbounds.Add(tag, &ci.wg)
+		if err != nil {
+			log.Printf("failed to add inbound: %v", err)
+		}
+		ci.inboundPorts[name] = inPort
+	}
+	for _, tag := range c.outboundStrings {
+		ci.outbounds.Add(tag)
+	}
+
+	return ci
+}
+
+func (ci *SshClientInstance) Start() error {
+	if !ci.k8sPhase {
+		// Attestation Phase
+		log.Println("Attstation Phase: Starting")
+		if err := ci.StartAttestation(); err != nil {
+			return fmt.Errorf("attstation Phase failed: %v", err)
+		}
+		log.Println("Attstation Phase: Done")
+		ci.k8sPhase = true
+	}
+
+	// Kubernetes Phase
+	ci.wg.Add(1)
+	go func() {
+		defer ci.wg.Done()
+		restarts := 0
+		for {
+			select {
+			case <-ci.ctx.Done():
+				log.Printf("Kubernetes Phase: Done")
+				return
+			default:
+				log.Printf("Kubernetes Phase: Starting (Number of restarts %d)", restarts)
+				if err := ci.StartKubernetes(); err != nil {
+					log.Printf("Kubernetes Phase: failed: %v", err)
+				}
+				time.Sleep(time.Second)
+				restarts += 1
+			}
+		}
+	}()
+	return nil
+}
+
+func (ci *SshClientInstance) StartKubernetes() error {
+	ctx, cancel := context.WithCancel(ci.ctx)
+	defer cancel()
+	peer := ci.StartSshClient(ctx, sshproxy.KUBERNETES, ci.publicKey, ci.sid)
+	if peer == nil {
+
+		return fmt.Errorf("Kubernetes Phase: failed StartSshClient")
+	}
+
+	peer.AddOutbounds(ci.outbounds)
+	err := peer.AddInbounds(ci.inbounds)
+	if err != nil {
+		peer.Close("Inbounds failed")
+		peer = nil
+		return fmt.Errorf("inbounds failed: %w", err)
+	}
+	peer.Wait()
+	return nil
+}
+
+func (ci *SshClientInstance) StartAttestation() error {
+	ctx, cancel := context.WithCancel(ci.ctx)
+	defer cancel()
+	peer := ci.StartSshClient(ctx, sshproxy.ATTESTATION, nil, ci.sid)
+	if peer == nil {
+		return fmt.Errorf("Attestation Phase: failed StartSshClient")
```

Comment:
```suggestion
		return fmt.Errorf("Attestation phase: failed StartSshClient")
```

It's weird that we don't get an error object from the `StartSshClient` in that case.